rspace/src/main/scala/coop/rchain/rspace/HotStore.scala

```diff
@@ -34,277 +28,372 @@ trait HotStore[F[_], C, P, A, K] {
-                      }
-                } yield (historyJoins)
-              case Some(joins) => Applicative[F].pure(joins)
-            }
-    } yield (res)
+      _              <- initHotJoins(channel)
+      joins          <- joinsRef.get
+      installedJoins <- instJoinsRef.get
+      r              = installedJoins.getOrElse(channel, Seq.empty[Seq[C]]) ++: joins(channel)
+    } yield r
 
   def putJoin(channel: C, join: Seq[C]): F[Unit] =
     for {
-      joins <- getJoins(channel)
-      _ <- Applicative[F].whenA(!joins.contains(join))(S.update { cache =>
-            ignore(cache.joins.put(channel, join +: joins))
-            cache
-          })
+      _ <- initHotJoins(channel)
+      _ <- joinsRef.update { state =>
+            val curVal = state(channel)
+            val newVal = join +: curVal
+            state.updated(channel, newVal)
+          }
     } yield ()
 
-  def installJoin(channel: C, join: Seq[C]): F[Unit] = S.update { cache =>
-    ignore {
-      val installed = cache.installedJoins.get(channel).getOrElse(Seq.empty)
-
-      if (!installed.contains(join))
-        cache.installedJoins
-          .put(channel, join +: installed)
-    }
-    cache
+  def installJoin(channel: C, join: Seq[C]): F[Unit] = instJoinsRef.update { state =>
+    val curInstalled = state.getOrElse(channel, Seq.empty)
+    if (!curInstalled.contains(join)) {
+      val newVal = join +: curInstalled
+      state.updated(channel, newVal)
+    } else
+      state
   }
 
   def removeJoin(channel: C, join: Seq[C]): F[Unit] =
     for {
-      joins         <- internalGetJoins(channel)
-      continuations <- getContinuations(join)
-      s             <- S.get
-      index         = s.joins(channel).indexOf(join)
-      _ <- if (continuations.isEmpty && index != -1)
-            checkIndex(s.joins(channel), index) >>
-              S.update { cache =>
-                val updated =
-                  removeIndex(
-                    cache.joins(
-                      channel
-                    ),
-                    index
-                  )
-                ignore(
-                  cache.joins.put(
-                    channel,
-                    updated
-                  )
-                )
-                cache
-              } else Applicative[F].unit
+      _              <- initHotJoins(channel)
+      _              <- initHotContinuations(join)
+      contsAtJoin    <- contRef.get.map(_.get(join))
+      instContAtJoin <- instContRef.get.map(_.get(join))
+      conts          = contsAtJoin ++ instContAtJoin
+      err <- joinsRef.modify[Boolean](state => {
+              val curJoins = state(channel)
+              val curConts = conts
+
+              val index       = state(channel).indexOf(join)
+              val outOfBounds = !curJoins.isDefinedAt(index)
+
+              // TODO should attimpting to remove join with non empty contnuation lead to error as well?
+              val doRemove = curConts.isEmpty
```

Comment:
Hm, this looks like a race. Reading from two Refs and depending on a value in the third. Two threads can enter this function, read the same continuations and result in different state.

Check to remove join depends on one version of the state (when continuations are read) so here when checking if it's empty it can give inconsistent results. Maybe we cannot escape to have on Ref for the whole state. Or maybe something similar to RSpace per channel locking.

Reply:
I have been thinking on this. Installed continuations should be filled only once when runtime is spawned, so `instContAtJoin` should never change here. And `conts` which is `contsAtJoin ++ instContAtJoin` - sum of this two non atomic reads, is used to detect if code should proceed and delete. So only race here cna be at `contsAtJoin` and can lead to attempting to remove the same join several times. I have been thinking maybe we can just check if join is not present anymore in the state and not throw an exception in that case but just don't do remove.

Reply:
https://github.com/rchain/rchain/pull/3271/commits/ce22b4e74f292374d7b5c2736896cc2b927b10e1#diff-84d533e963703b6bf252ff208084a62af1116772648322604b3869ad15b8c39cR224